source/common/router/config_utility.cc

```diff
@@ -38,16 +38,19 @@ ConfigUtility::QueryParameterMatcher::QueryParameterMatcher(
     : name_(config.name()), matcher_(maybeCreateStringMatcher(config)) {}
 
 bool ConfigUtility::QueryParameterMatcher::matches(
-    const Http::Utility::QueryParams& request_query_params) const {
-  auto query_param = request_query_params.find(name_);
-  if (query_param == request_query_params.end()) {
+    const Http::Utility::QueryParamsMulti& request_query_params) const {
+  // The docs specify that only the first instance of the query parameter should be matched
```

Comment:
Ah - I found it [here](https://www.envoyproxy.io/docs/envoy/latest/api-v3/type/matcher/v3/http_inputs.proto#type-matcher-v3-httprequestqueryparammatchinput).  Now that I'm staring at it, that might be a different query parameter matcher.

Should we update the docs you linked, or introduce a behavior change?

Reply:
I did some reading on it, and it seems the consensus is that it's loosely-specified at best, and really implementation-dependent, and some implementations depend on order.

But in this case, we're talking about repeated params, eg `?p=1&p=2&p=3`, which is also loosely specified, and more likely that the implementations depend on order. So order-preserving makes sense here.

I think this PR should have no behavior change as-written. If we need more tests to validate them, then let's add them.

The context from this is that previously, the ext_authz filter would in some configurations discard all-but-one of a repeated query param. @turettn wrote a better API for query params that handles repeated properly, and I asked to use that as a replacement for the old one entirely, so that's what this change is.

Reply:
Based on @ggreenway's comment, I'm tempted to leave the code as is and remove the erroneous comment.  If we want to make the behavior change, we can make that at a later date.

@alyssawilk Does that work for you?

Reply:
I'm going to pull in @DavidSchinazi and @RyanTheOptimist because I suspect they know more about proper url handling than I.

Reply:
URI query parameters are very loosely defined at the IETF layer (cf [RFC 3986 s3.4](https://www.rfc-editor.org/rfc/rfc3986.html#section-3.4)). At the W3C layer, there is some acknowledgement that repeated parameters are allowed and that ordering can carry semantic value (cf [definition of URLSearchParams](https://url.spec.whatwg.org/#interface-urlsearchparams)). Therefore matching on the first one doesn't violate any of the specs. Given that someone might be relying on this behavior, I would keep the behavior as-is and update the docs/comments to match the behavior of the code

Reply:
Yep, @DavidSchinazi that was my understanding as well.

But regardless, this PR is trying to maintain exactly the previous Envoy behavior. It may be worth adding a different mode later when handling repeated parameters, but I think this PR is not a behavior change.